### DISCUSSION TITLE
ShowDialog with Callback or ShowDialogAsync?

### DISCUSSION THREAD

**Discussion by Shannon-roos** [PrismLibrary/Prism#3282] [.NET MAUI] [answered]
Now that the dialogservice is available for .NET MAUI in v9 I've noticed there is a remark on the .ShowDialog method saying "<remarks>This is for backwards compatibility. Use DialogCallback instead.</remarks>"

I am migrating from Xaamarin Forms where I was using the 

The question is, should I migrate to use the ShowDialogAsync method and just return the IDialogResult and deal with it without a callback? Or, should I migrate to the "new?" DialogCallback for which I can't seem to find any examples or doccumentation about? 

For completeness I am using an extension class and most of the dialogs are confirmation or Alert Dialogs (Yers\No, Alert Message, Enter Value etc) that need to be modal\blocking. 

I feel like ShowDialogAsync is nicer but wanted to ask if thats the best case moving forward?

**Comment by dansiegel** [comment]
So the answer is really it depends on your needs. It sounds like you want:

```cs
var result = await dialogs.ShowDialogAsync("foo");
if (result.Parameters.GetValue<bool>("bar"))
{
    // do something
}
```

If that's what you need then that's probably the best API to use. But on the other hand you could potentially do this too... it really all depends on your needs.

```cs
private void OnSomeCommandExecuted()
{
    dialogs.ShowDialog("foo", DialogCallback.OnCloseAsync(async result => {
        if (result.Parameters.GetValue<bool>("bar"))
            await DoSomething();
    });
}
```

**Reply by dansiegel** [reply, marked as answer]
btw the samples are in the docs:
https://docs.prismlibrary.com/docs/dialogs/index.html#on-close

**Comment by Shannon-roos** [comment]
Thanks Dan for the clarifications. 

Considering I was coming from the "make it async using the TaskCompletionSource" approach,  the Async method makes more sense for my use case. Kind of feel a bit funny looking at it now. 😄 (The Tuple also really sets it off!)  

```
//This isn't even an extension!
 public Task<(eDialogReturnResult DialogResult, Trader selectedTrader)> ShowSelectTraderDialog(
                    string message,
                    eTraderType trdaerType)
        {
            _logger.Debug($"Message:{message}");
            var tcs = new TaskCompletionSource<(eDialogReturnResult DialogResult, Trader selectedTrader)>();
            try
            {
                DialogParameters p = new DialogParameters();
                p.Add("message", message);
                p.Add("traderType", trdaerType);
             
                _dialogService.ShowDialog("SelectTraderDialogView", p, (dparams) =>
                {

                    eDialogReturnResult res = dparams.Parameters.GetValue<eDialogReturnResult>("result");
                    Trader retTrader = dparams.Parameters.GetValue<Trader>("trader");

                    tcs.SetResult((res, retTrader));


                });
            }
            catch (Exception ex)
            {
                _logger.Error($"Exception Loading Dialog:{ex}");
                tcs.TrySetException(ex);
            }

            return tcs.Task;
        }
```


I do like the DialogCallback approach as well. I must've been having a moment becasue I couldn't figure out how to use that in an Extension method from he docs but your second code block made it all make sense. 

Cheers again.
